src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,911 @@
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => {
+                declare_variable(name, val, env);
+                succeed("variable declaration ok", fail2);
+            },
+            fail);
+}
+
+function analyze_constant_declaration(stmt) {
+    const name =
+        constant_declaration_name(stmt);
+    const vfun =
+        analyze(constant_declaration_value(stmt));
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => {
+                declare_constant(name, val, env);
+                succeed("constant declaration ok", fail2);
+            },
+            fail);
+}
+
+function analyze_assignment(stmt) {
+    const variable = assignment_name(stmt);
+    const vfun = analyze(assignment_value(stmt));
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => { // *1*
+                const old_value = lookup_name_value(variable, env);
+                set_name_value(variable, val, env);
+                succeed("assignment ok",
+                    () => { // *2*
+                        set_name_value(variable, old_value, env);
+                        fail2();
+                    });
+            },
+            fail);
+}
+
+function analyze_application(stmt) {
+    const ffun = analyze(operator(stmt));
+    const afuns = map(analyze, operands(stmt));
+    return (env, succeed, fail) =>
+        ffun(env,
+            (fun, fail2) =>
+            get_args(afuns,
+                env,
+                (args, fail3) =>
+                execute_application(fun,
+                    args, succeed, fail3),
+                fail2),
+            fail);
+}
+
+function get_args(afuns, env, succeed, fail) {
+    return is_null(afuns) ?
+        succeed(null, fail) :
+        head(afuns)(env,
+            // success continuation for this afun
+            (arg, fail2) =>
+            get_args(tail(afuns),
+                env,
+                // success continuation for
+                // recursive call to get_args
+                (args, fail3) =>
+                succeed(pair(arg, args),
+                    fail3),
+                fail2),
+            fail);
+}
+
+function execute_application(fun, args, succeed, fail) {
+    if (is_primitive_function(fun)) {
+        succeed(apply_primitive_function(fun, args), fail);
+    } else if (is_compound_function(fun)) {
+        const body = function_body(fun);
+        const locals = function_locals(fun);
+        const names = insert_all(map(name_of_name, function_parameters(fun)), locals);
+        const temp_values = map(x => no_value_yet, locals);
+        const values = append(args, temp_values);
+
+        const application_value = function_body(fun)(
+            extend_environment(
+                names,
+                values,
+                function_environment(fun)),
+            succeed,
+            fail);
+        if (is_return_value(application_value)) {
```

Comment:
(Please correct me if I'm wrong here), but I think this logic should be placed within the success continuation that is given to the `function_body(fun)` execution function?

Currently, if we try:
```
parse_and_run("function foo() {\
    5;\
}\
foo();");
```

the output is:
```
"// Amb-Eval value:"
5
```

Instead, it should be:
```
"// Amb-Eval value:"
undefined
```

This could be solved by placing the `if-else` logic within the success continuation given to the function body's execution function. Just a suggestion to consider!

Reply:
Thank you! Updated at 547be0b